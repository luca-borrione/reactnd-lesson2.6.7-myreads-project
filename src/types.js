
import PropTypes from 'prop-types'

/**
 * @typedef {Object} TBook
 * @description
 * Custom type to define a book internally.
 * It's filled with properties coming from the
 * book literal object fetched using the BooksAPI
 * @property {string[]} [authors] - List of the authors names
 * @property {string} id - The book unique id
 * @property {string} [shelf] - The shelf unique name
 * @property {string} [thumbnail] - URL for the cover preview image
 * @property {string} title - The book's title
 */

export const TBook = {
	authors: PropTypes.arrayOf(PropTypes.string.isRequired),
	id: PropTypes.string.isRequired,
	shelf: PropTypes.string,
	thumbnail: PropTypes.string,
	title: PropTypes.string.isRequired
};

