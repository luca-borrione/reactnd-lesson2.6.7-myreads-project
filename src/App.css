html, body, .root {
  height: 100%;
}
body {
  line-height: 1.5;
}
body, .app {
  background: white;
}

/* main page */

.list-books-title {
  padding: 10px 0;
  background: #2e7c31;
  text-align: center;
}
.list-books-title h1 {
  font-weight: 400;
  margin: 0;
  color: white;
}

.list-books-content {
  padding: 0 0 80px;
  flex: 1;
}

.bookshelf {
  padding: 0 10px 20px;
}

@media (min-width: 600px) {
  .bookshelf {
    padding: 0 20px 40px;
  }
}

.bookshelf-title {
  border-bottom: 1px solid #dedede;
}
.bookshelf-books {
  text-align: center;
}

.open-search {
  position: fixed;
  right: 25px;
  bottom: 25px;
}
.open-search a {
  display: block;
  width: 50px;
  height: 50px;
  border-radius: 50%;
  background: #2e7d32;
  background-image: url('./icons/add.svg');
  background-repeat: no-repeat;
  background-position: center;
  background-size: 28px;
  box-shadow: 0 3px 6px rgba(0,0,0,0.16), 0 3px 6px rgba(0,0,0,0.23);
  font-size: 0;
}

/* search page */

.search-books-bar {
  position: fixed;
  width: 100%;
  top: 0;
  left: 0;
  z-index: 5;
  display: flex;
  box-shadow: 0 10px 20px rgba(0,0,0,0.19), 0 0 6px rgba(0,0,0,0.23);
}
.search-books-input-wrapper {
  flex: 1;
  background: #e9e;
}
.search-books-bar input {
  width: 100%;
  padding: 15px 10px;
  font-size: 1.25em;
  border: none;
  outline: none;
}

.close-search {
  display: block;
  top: 20px;
  left: 15px;
  width: 50px;
  height: 53px;
  background: white;
  background-image: url('./icons/arrow-back.svg');
  background-position: center;
  background-repeat: no-repeat;
  background-size: 28px;
  font-size: 0;
}

.search-books-results {
  padding: 80px 10px 20px;
}

/* books grid */

.books-grid {
  list-style-type: none;
  padding: 0;
  margin: 0;

  display: flex;
  justify-content: center;
  flex-wrap: wrap;
}
.books-grid li {
  padding: 10px 15px;
  text-align: left;
}

.book {
  width: 140px;
}
.book-title,
.book-authors {
  font-size: 0.8em;
}
.book-title {
  margin-top: 10px;
}
.book-authors {
  color: #999;
}
.book-authors .author {
	display: block;
}

.book-top {
  position: relative;
  height: 200px;
  display: flex;
  align-items: flex-end;
}

.book-shelf-changer {
  position: absolute;
  right: 0;
  bottom: -10px;
  width: 40px;
  height: 40px;
  border-radius: 50%;
  background: #60ac5d;
  background-image: url('./icons/arrow-drop-down.svg');
  background-repeat: no-repeat;
  background-position: center;
  background-size: 20px;
  box-shadow: 0 3px 6px rgba(0,0,0,0.16), 0 3px 6px rgba(0,0,0,0.23);
}
.book-shelf-changer select {
  width: 100%;
  height: 100%;
  opacity: 0;
  cursor: pointer;
}

/* book cover */

.book-cover {
  box-shadow: 0 3px 6px rgba(0,0,0,0.16), 0 3px 6px rgba(0,0,0,0.23);
  background: #eee;
}
.book-cover-title {
  padding: 20px 10px 0;
  text-align: center;
  font-size: 0.8em;
}

/* messaging */
.panel-info, .panel-success, .panel-warning, .panel-error, .panel-validation {
  border: 0.0625em solid;
  margin: 0.625em 0;
  padding: 0.9375em 0.625em 0.9375em 3.125em;
  background-repeat: no-repeat;
  background-position: 0.625em center;
  z-index: 1000;
}
.search-books [class|="panel"] {
  position: relative;
  top: 3.3125em;
}
.panel-info {
  color: #00529B;
  background-color: #BDE5F8;
  background-image: url('https://i.imgur.com/ilgqWuX.png');
}
.panel-success {
  color: #4F8A10;
  background-color: #DFF2BF;
  background-image: url('https://i.imgur.com/Q9BGTuy.png');
}
.panel-warning {
  color: #9F6000;
  background-color: #FEEFB3;
  background-image: url('https://i.imgur.com/Z8q7ww7.png');
}
.panel-error{
  color: #D8000C;
  background-color: #FFBABA;
  background-image: url('https://i.imgur.com/GnyDvKN.png');
}
.panel-validation{
  color: #D63301;
  background-color: #FFCCBA;
  background-image: url('https://i.imgur.com/GnyDvKN.png');
}

/* bookshelf changer loading */
#loading {
  display: inline-block;
  width: 2.5em;
  height: 2.5em;
  border: 0.1875em solid rgba(255,255,255,.3);
  border-radius: 50%;
  border-top-color: #fff;
  animation: spin 1s linear infinite;
  -webkit-animation: spin 1s linear infinite;
  -moz-animation: spin 1s linear infinite;
  -o-animation: spin 1s linear infinite;
  -ms-animation: spin 1s linear infinite;
  animation: spin 1s linear infinite;
}

@-webkit-keyframes spin {
  to { -webkit-transform: rotate(360deg); }
}
@-moz-keyframes spin {
  to { -moz-transform: rotate(360deg); }
}
@-o-keyframes spin {
  to { -o-transform: rotate(360deg); }
}
@-ms-keyframes spin {
  to { -ms-transform: rotate(360deg); }
}
@keyframes spin {
  to { transform: rotate(360deg); }
}


/* book loader */
.book-loader {
  position: fixed;
  left: 50%;
  top: 50%;
  transform: translateY(-50%);
  border: 0.3125em solid #205722;
  background: #3ea842;
  width: 6.25em;
  margin-left: -3.125em;
  height: 3.75em;
  opacity: 0;
  animation: fadeIn 0.5s 1 ease-in;
  animation-fill-mode: forwards;
  animation-delay: 1s;
}
@-webkit-keyframes fadeIn {
	0% { opacity: 0;}
	100% { opacity: 1; }
}
@-moz-keyframes fadeIn {
	0% { opacity: 0;}
	100% { opacity: 1; }
}
@-o-keyframes fadeIn {
	0% { opacity: 0;}
	100% { opacity: 1; }
}
@-ms-keyframes fadeIn {
	0% { opacity: 0;}
	100% { opacity: 1; }
}
@keyframes fadeIn {
	0% { opacity: 0;}
	100% { opacity: 1; }
}
.book-page {
  position: absolute;
  left: 50%;
  top: -0.3125em;
  margin: 0 auto;
  border-top: 0.3125em solid #205722;
  border-bottom: 0.3125em solid #205722;
  border-right: 0.3125em solid #205722;
  background: #c7eac8;
  width: 3.125em;
  height: 3.75em;
  transform-origin: 0% 50%;
  animation: flip 1.2s infinite linear;
  animation-fill-mode: forwards;
}
.book-page:nth-child(1) {
  z-index: -1;
  animation-delay: 1.4s;
}
.book-page:nth-child(2) {
  z-index: -2;
  animation-delay: 2.8s;
}
.book-page:nth-child(3) {
  z-index: -3;
  animation-delay: 4.2s;
}
@-webkit-keyframes flip {
  0% {
    -webkit-transform: perspective(600px) rotateY(0deg);
  }
  20% {
    background: #205722;
  }
  29.9% {
    background: #205722;
  }
  30% {
    -webkit-transform: perspective(200px) rotateY(-90deg);
    background: #2e7c31;
  }
  54.999% {
    opacity: 1;
  }
  55% {
    opacity: 0;
  }
  60% {
    -webkit-transform: perspective(200px) rotateY(-180deg);
    background: #2e7c31;
  }
  100% {
    -webkit-transform: perspective(200px) rotateY(-180deg);
    background: #2e7c31;
  }
}
@-moz-keyframes flip {
  0% {
    -moz-transform: perspective(600px) rotateY(0deg);
  }
  20% {
    background: #205722;
  }
  29.9% {
    background: #205722;
  }
  30% {
    -moz-transform: perspective(200px) rotateY(-90deg);
    background: #2e7c31;
  }
  54.999% {
    opacity: 1;
  }
  55% {
    opacity: 0;
  }
  60% {
    -moz-transform: perspective(200px) rotateY(-180deg);
    background: #2e7c31;
  }
  100% {
    -moz-transform: perspective(200px) rotateY(-180deg);
    background: #2e7c31;
  }
}
@-o-keyframes flip {
  0% {
    -o-transform: perspective(600px) rotateY(0deg);
  }
  20% {
    background: #205722;
  }
  29.9% {
    background: #205722;
  }
  30% {
    -o-transform: perspective(200px) rotateY(-90deg);
    background: #2e7c31;
  }
  54.999% {
    opacity: 1;
  }
  55% {
    opacity: 0;
  }
  60% {
    -o-transform: perspective(200px) rotateY(-180deg);
    background: #2e7c31;
  }
  100% {
    -o-transform: perspective(200px) rotateY(-180deg);
    background: #2e7c31;
  }
}
@-ms-keyframes flip {
  0% {
    -ms-transform: perspective(600px) rotateY(0deg);
  }
  20% {
    background: #205722;
  }
  29.9% {
    background: #205722;
  }
  30% {
    -ms-transform: perspective(200px) rotateY(-90deg);
    background: #2e7c31;
  }
  54.999% {
    opacity: 1;
  }
  55% {
    opacity: 0;
  }
  60% {
    -ms-transform: perspective(200px) rotateY(-180deg);
    background: #2e7c31;
  }
  100% {
    -ms-transform: perspective(200px) rotateY(-180deg);
    background: #2e7c31;
  }
}
@keyframes flip {
  0% {
    transform: perspective(600px) rotateY(0deg);
  }
  20% {
    background: #205722;
  }
  29.9% {
    background: #205722;
  }
  30% {
    transform: perspective(200px) rotateY(-90deg);
    background: #2e7c31;
  }
  54.999% {
    opacity: 1;
  }
  55% {
    opacity: 0;
  }
  60% {
    transform: perspective(200px) rotateY(-180deg);
    background: #2e7c31;
  }
  100% {
    transform: perspective(200px) rotateY(-180deg);
    background: #2e7c31;
  }
}


/* 404 not found */
.notfound-page-wrap {
  width: 400px;
  height: 400px;
  border-radius: 50%;
  background: #2e7c31;
  position: absolute;
  transform: translate(-50%, -50%);
  left: 50%;
  top: 50%;
  text-align: Center;
  display: flex;
  flex-direction: column;
  justify-content: center;
}
.notfound-page-wrap h1 {
  font-size: 100px;
  margin: 0;
  color: #fff;
}
.notfound-page-wrap h2 {
  margin: 0;
  color: #fff;
  margin: 20px 0;
}
.notfound-page-wrap a {
  color: #fff;
  display: block;
}