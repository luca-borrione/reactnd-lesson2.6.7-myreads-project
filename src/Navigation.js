import React from 'react';
import PropTypes from 'prop-types';
import { Route, Switch } from 'react-router-dom';
import PropsRoute from './PropsRoute';
import BooksList from './BooksList';
import SearchPage from './SearchPage';
import NotFoundPage from './NotFoundPage';
import { TBook } from './types';


const Navigation = props => {

	const { booksInShelves, getBookShelf, updateBookShelf} = props;

	return (
		<Switch>
			<PropsRoute exact path='/' component={BooksList}
				booksInShelves={booksInShelves}
				updateBookShelf={updateBookShelf} />

			<PropsRoute path='/search' component={SearchPage}
				getBookShelf={getBookShelf}
				updateBookShelf={updateBookShelf} />

			<Route component={NotFoundPage} />
		</Switch>
	);

};

Navigation.propTypes = {
	booksInShelves:	PropTypes.arrayOf(
		PropTypes.shape(TBook).isRequired
	).isRequired,
	getBookShelf: PropTypes.func.isRequired,
	updateBookShelf: PropTypes.func.isRequired
};

export default Navigation;
