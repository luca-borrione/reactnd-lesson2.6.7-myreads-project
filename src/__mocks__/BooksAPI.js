import * as data from './allBooks.json'
import { TShelfKey } from '../shared/types'

const DELAY = 10;

const getDelay = () => DELAY * (Math.floor(Math.random() * 10) + 1); // random between 1 - 10

export const MOCKED = true;

let allBooks;
export const __resetBooks = () => {
	allBooks = JSON.parse(JSON.stringify(data.books)); // cloning
};
__resetBooks();

const getBooksInShelves = () =>
	allBooks.filter( ({ shelf }) => shelf !== TShelfKey.NONE );


export const getAll = () => {
	const booksInShelves = getBooksInShelves();
	return new Promise( resolve => {
		const delay = getDelay();
		setTimeout(() => {
			resolve(booksInShelves);
		}, delay);
	});
};

export const get = bookId => {
	const book = allBooks.find( ({ id }) => id === bookId );
	return new Promise( resolve => {
		const delay = getDelay();
		setTimeout(() => {
			resolve(book);
		}, delay);
	});
};

const formatUpdatedBooks = (books) => {
	const fBooks = {};

	Object.keys(TShelfKey).forEach( key => {
		const shelfKey = TShelfKey[key];
		if (shelfKey !== TShelfKey.NONE) {
			fBooks[shelfKey] = books
							.filter( ({ shelf }) => shelf === shelfKey )
							.map( book => book.id );
		}
	});

	return fBooks
};

export const update = (book, shelf) => {
	const existingBook = allBooks.find( ({ id }) => id === book.id );
	existingBook.shelf = shelf;
	const booksInShelves = getBooksInShelves();
	return new Promise( resolve => {
		const delay = getDelay();
		setTimeout(() => {
			resolve(formatUpdatedBooks(booksInShelves));
		}, delay);
	});
};

export const search = (query, maxResults) => {
	const re = new RegExp(`^${query}`, 'i');
	const result = allBooks.filter( ({ title, authors }) => {
		return re.test(title);
	})
	.slice(0, maxResults);

	return new Promise( resolve => {
		const delay = getDelay();
		setTimeout(() => {
			resolve(result);
		}, delay);
	});
};

