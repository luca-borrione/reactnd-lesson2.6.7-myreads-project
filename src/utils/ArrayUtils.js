
export const diff = (...arrays) => {
	return [].concat(...arrays.map( (arr, i) => {
		const others = arrays.slice(0);
		others.splice(i, 1);
		const unique = [...new Set([].concat(...others))];
		return arr.filter(x => !unique.includes(x));
	}));
};

export const diffByKey = (key, ...arrays) => {
	return [].concat(...arrays.map( (arr, i) => {
		const others = arrays.slice(0);
		others.splice(i, 1);
		const unique = [...new Set([].concat(...others))];
		return	arr.filter( x =>
					unique.filter( y =>
						x[key] === y[key]).length === 0 );
	}));
}