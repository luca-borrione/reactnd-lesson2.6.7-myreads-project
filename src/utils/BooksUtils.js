
import { makeTitle } from './StringUtils'

export const getShelves = books => {
	return [
		...new Set(books.map( book => book.shelf ))
	];
};

export const getShelfTitle = shelf => {
	let title;

	switch (shelf) {

		case 'currentlyReading':
			title = 'currently reading';
			break;

		case 'wantToRead':
			title = 'want to read';
			break;

		case 'read':
			title = 'read';
			break;

		default:
			throw new Error('unexpected shelf id: '+shelf);

	}

	return makeTitle(title);
};

export const getBooksInShelf = (books, shelf) => {
	return books.filter( book => book.shelf === shelf );
}