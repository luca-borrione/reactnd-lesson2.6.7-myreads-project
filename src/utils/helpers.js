import settings from '../settings';

export function getBasename(): string {
  const basename: {
    [key: string]: string
  } = settings.basename || {};
  return (basename[window.location.hostname])
    ? `/${basename[window.location.hostname]}/`
    : '/';
}

export function addBasenameToUrl(url: string): string {
  const basename: string = getBasename();
  return (url.indexOf(basename) === -1)
    ? basename + url.replace(/^\/+/g, '')
    : url;
}
