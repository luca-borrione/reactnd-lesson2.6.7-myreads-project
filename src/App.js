import React from 'react';
import Navigation from './Navigation';
import BookLoader from './BookLoader';
import PanelError from './PanelError';
import * as BooksAPI from './BooksAPI';
import { TShelfKey } from './shared/types';
import './App.css';
import { ERROR } from './shared/constants';

/**
 * @module
 * @name App
 * @extends React.Component
 * @description
 * When first loaded it automatically fetches all the books in the shelves
 * and it stores them internally as a state.<br>
 * This becomes the source of truth across the different sections of the app.
 */
class App extends React.Component {

	/**
	 * @member
	 * @name STATUS
	 * @description Collection of possible status
	 * @property {string} INITIAL - The app is loading
	 * @property {string} BUSY - The app is remotely updating the books in the shelves
	 * @property {string} READY - Either all the books has been fetched initially
	 * 								or the books has been remotely updated
	 * @property {string} ERROR - An error occurred when tryng to fetch the books or to update them
	 * @static
	 */
	static STATUS = {
		INITIAL: 'STATUS.INITIAL',
		BUSY: 'STATUS.BUSY',
		READY: 'STATUS.READY',
		ERROR: 'STATUS.ERROR'
	};

	constructor(props) {
		super(props);
		this.updateBookShelf = this.updateBookShelf.bind(this);
		this.getBookShelf = this.getBookShelf.bind(this);
	}


	/**
	 * @member
	 * @name state
	 * @property {STATUS} status - one of the possible [STATUS]{@link module:App.STATUS}
	 * @property {TBook[]} booksInShelves - List of all the books present in all the shelves
	 * @private
	 */
	state = {
		status: this.constructor.STATUS.INITIAL,
		booksInShelves: null
	};


	/**
	 * @member
	 * @name async
	 * @description
	 * Collection of promises useful when making unit tests.
	 * @property {Promise} fetchAllBooks - waiting for the books in the shelves to be fetched
	 */
	async = {
		fetchAllBooks: null
	};


	componentDidMount() {
		this.fetchAllBooks();
	}


	/**
	 * @method
	 * @name fetchAllBooks
	 * @description
	 * Asynchronously retrieves the list of all the books currently in a shelf from remote,
	 * then it stores them internally in the [state]{@link module:App~state} as booksInShelves collection.
	 * @returns {Promise}
	 * @private
	 */
	fetchAllBooks() {
		return this.async.fetchAllBooks = new Promise( async resolve => {
			const { STATUS } = this.constructor;
			let booksInShelves = [];

			try {
				booksInShelves = await BooksAPI.getAll();

				// NOTE: uncomment to test the initial loader animation
				// await new Promise(resolve => {
				// 	setTimeout(()=>{ resolve(); }, 10000);
				// });

				await this.setState({
					status: STATUS.READY,
					booksInShelves
				});

			} catch (error) {
				console.error(error); // NOTE: this is intentional
				await this.setState({
					status: STATUS.ERROR,
					booksInShelves
				});
			}

			resolve(booksInShelves);
		});
	}


	/**
	 * @method
	 * @name updateBookShelf
	 * @description
	 * Updates the remote list of books in the shelves,
	 * then it changes the internal [booksInShelves]{@link module:App~state} collection in the state accordingly.
	 * @param {TBook} book - The book the user selected
	 * @param {TShelfKey} shelf - The shelf key the user selected
	 * @returns {TBook[]} booksInShelves
	 */
	async updateBookShelf(book, shelf) {
		const { STATUS } = this.constructor;
		const { booksInShelves: currBooksInShelves} = this.state;

		let booksInShelves = [];
		let status = STATUS.BUSY;

		await this.setState({ status });

		try {
			const bookIdsInShelves = await BooksAPI.update(book, shelf);

			switch (shelf) {
				// Removing a book
				case TShelfKey.NONE:
					// Checking that the API actually worked
					const updated = await BooksAPI.get(book.id);
					if (updated.shelf !== shelf) {
						throw new Error(`${ERROR.BOOKS_API_FAILED}: update(${book.id}, ${shelf})`);
					}

					booksInShelves = currBooksInShelves.filter( ({ id }) => id !== book.id );
					break;

				default:
					// Checking that the API actually worked
					if (!bookIdsInShelves[shelf].find( id => id === book.id)) {
						throw new Error(`${ERROR.BOOKS_API_FAILED}: update(${book.id}, ${shelf})`);
					}

					const bookIndex = currBooksInShelves.findIndex( ({ id }) => id === book.id );

					if (bookIndex === -1) {
						// Adding a new book
						book.shelf = shelf;
						booksInShelves = [...currBooksInShelves, book];
					} else {
						// Moving an existing book to a new shelf
						currBooksInShelves[bookIndex].shelf = shelf;
						booksInShelves = currBooksInShelves;
					}
					break;
			}

			status = STATUS.READY;

		} catch (error) {
			console.error(error);
			status = STATUS.ERROR;
		}

		await this.setState({ status, booksInShelves });

		return booksInShelves;
	}


	/**
	 * @method
	 * @name getBookShelf
	 * @description
	 * Retrieves the current shelf associated with a book, based on a given book id.
	 * If the book is not present in the [booksInShelves]{@link module:App~state} collection
	 * the shelf will be NONE.
	 * @param {string} bookID - The book id as described in [TBook.id]{@link module:types.TBook}
	 * @returns {TShelfKey}
	 */
	getBookShelf(bookID) {
		const { booksInShelves } = this.state;
		const book = booksInShelves.find( ({id}) => id === bookID );
		return book ? book.shelf : TShelfKey.NONE;
	}


	/**
	 * @method
	 * @name shouldComponentUpdate
	 * @param {Object} nextProps - ignored as the module does not expect props
	 * @param {state} nextState - refers to [state]{@link module:App~state}
	 * @returns {boolean} - false if status is STATUS.BUSY
	 * @private
	 */
	shouldComponentUpdate(nextProps, nextState) {
		const { STATUS } = this.constructor;
		return nextState.status !== STATUS.BUSY;
	}


	/**
	 * @method
	 * @name render
	 * @description
	 * Renders
	 * - [BookLoader]{@link module:BookLoader} when the status is STATUS.INITIAL
	 * - [Navigation]{@link module:Navigation} when the status is STATUS.READY
	 * - [PanelError]{@link module:PanelError} when the status is STATUS.ERROR
	 * @throws Error for an unexpected status
	 * @returns {ReactElement}
	 * @private
	 */
	render() {
		const { STATUS } = this.constructor;
		const { booksInShelves, status } = this.state;

		switch (status) {
			case STATUS.INITIAL:
				return <BookLoader />;

			case STATUS.ERROR:
				return <PanelError />;

			case STATUS.READY:
				return <Navigation
							booksInShelves={booksInShelves}
							getBookShelf={this.getBookShelf}
							updateBookShelf={this.updateBookShelf} />;

			default:
				throw new Error(ERROR.UNEXPECTED_STATUS, status);
		}
	}
}

export default App;
