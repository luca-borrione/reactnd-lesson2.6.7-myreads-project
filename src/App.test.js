import ReactDOM from 'react-dom';
import { mount, shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router';
import App from './App';
import Navigation from './Navigation';
import * as BooksAPI from './BooksAPI'; // mocked
import { diffByKey as ArrayDiffByKey } from './utils/ArrayUtils';
import { TShelfKey } from './types';

describe('App', () => {

	beforeEach(() => {
		jest.resetModules();
		BooksAPI.__resetBooks();
	});

	afterEach(() => {
		jest.clearAllMocks();
		jest.restoreAllMocks();
	});


	it('triggers componentDidMount and fetchAllBooks when mounted', () => {
		const componentDidMount = jest.spyOn(App.prototype, 'componentDidMount');
		const fetchAllBooks = jest.spyOn(App.prototype, 'fetchAllBooks');
		shallow(<App />);
		expect(componentDidMount).toHaveBeenCalledTimes(1);
		expect(fetchAllBooks).toHaveBeenCalledTimes(1);
	});


	it('renders without crashing', () => {
		const div = document.createElement('div');
		let component;
		ReactDOM.render(
			<MemoryRouter>
				<App ref={c => component = c} />
			</MemoryRouter>, div,
			async () => {
				await component.async.fetchAllBooks;
				ReactDOM.unmountComponentAtNode(div);
			});
	});


	it('renders correctly', () => {
		const tree = renderer.create(
			<MemoryRouter>
				<App />
			</MemoryRouter>).toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('should contain Navigation as a child', () => {
		const wrapper = mount(
			<MemoryRouter>
				<App />
			</MemoryRouter>);
		expect(wrapper.find(Navigation)).toBeTruthy();
	});


	describe('state behaviour', () => {

		it('fetches the books correctly and stores them as state', async () => {
			const expected = {
				books: await BooksAPI.getAll()
			};
			const wrapper = shallow(<App />);
			const component = wrapper.instance();
			await component.async.fetchAllBooks;

			const diffArr = ArrayDiffByKey('id', expected.books, wrapper.state().booksInShelves);
			expect(diffArr).toHaveLength(0);
		});


		it('properly removes a book and updates the state accordingly', async () => {
			const wrapper = shallow(<App />);
			const component = wrapper.instance();
			await component.async.fetchAllBooks;
			const initialBooksInShelves = wrapper.state().booksInShelves;

			const book = initialBooksInShelves[0]; // will remove the first book
			await component.updateBookShelf(book, TShelfKey.NONE);
			const booksInShelves = wrapper.state().booksInShelves;

			expect.assertions(3);
			{
				// 1. One book has ben removed
				expect(booksInShelves).toHaveLength(initialBooksInShelves.length - 1);

				// 2. The expected book has been removed
				const diffArr = ArrayDiffByKey('id', booksInShelves, initialBooksInShelves);
				const removedBook = diffArr[0];
				const expected = {
					id: book.id
				};
				expect(removedBook).toEqual(expect.objectContaining(expected));
			}
			{
				// 3. The removed book is now stored as NOT associated with any shelf
				const removedBook = await BooksAPI.get(book.id);
				const expected = {
					id: book.id,
					shelf: TShelfKey.NONE
				};
				expect(removedBook).toEqual(expect.objectContaining(expected));
			}
		});


		it('properly adds a new book and updates the state accordingly', async () => {
			const wrapper = shallow(<App />);
			const component = wrapper.instance();
			await component.async.fetchAllBooks;
			const initialBooksInShelves = wrapper.state().booksInShelves;

			// A new book is added
			const expected = {
				id: '1OJ8EhvuPXAC',
				shelf: TShelfKey.WANT_TO_READ // Moving from NONE
			};
			const book = await BooksAPI.get(expected.id);
			await component.updateBookShelf(book, expected.shelf);
			const booksInShelves = wrapper.state().booksInShelves;

			expect.assertions(3);
			{
				// 1. One book has ben added
				expect(booksInShelves).toHaveLength(initialBooksInShelves.length + 1);

				// 2. The expected book has been added
				const diffArr = ArrayDiffByKey('id', booksInShelves, initialBooksInShelves);
				const addedBook = diffArr[0];
				expect(addedBook).toEqual(expect.objectContaining(expected));
			}
			{
				// 3. The added book is now stored as expected
				const addedBook = await BooksAPI.get(expected.id);
				expect(addedBook).toEqual(expect.objectContaining(expected));
			}
		});


		it('properly moves a book to a different shelf and updates the state accordingly', async () => {
			const wrapper = shallow(<App />);
			const component = wrapper.instance();
			await component.async.fetchAllBooks;
			const initialBooksInShelves = wrapper.state().booksInShelves;

			// A book needs to move to a different shelf
			const expected = {
				id: 'nggnmAEACAAJ',
				shelf: TShelfKey.READ // moving from CURRENTLY_READING
			};
			const book = await BooksAPI.get(expected.id);
			await component.updateBookShelf(book, expected.shelf);
			const booksInShelves = wrapper.state().booksInShelves;

			expect.assertions(2);
			{
				// 1. No book has been added or removed
				expect(booksInShelves).toHaveLength(initialBooksInShelves.length);
			}
			{
				// 2. The expected book has been moved
				const movedBook = await BooksAPI.get(expected.id);
				expect(movedBook).toEqual(expect.objectContaining(expected));
			}
		});

	});


	it('retrieves the correct shelf associated with a given book', async () => {
		const wrapper = shallow(<App />);
		const component = wrapper.instance();
		await component.async.fetchAllBooks;

		const expectedArr = [
			{ id:'nggnmAEACAAJ', shelf:TShelfKey.CURRENTLY_READING },
			{ id:'evuwdDLfAyYC', shelf:TShelfKey.WANT_TO_READ },
			{ id:'jAUODAAAQBAJ', shelf:TShelfKey.READ },
			{ id:'tXrPCgAAQBAJ', shelf:TShelfKey.NONE }
		];

		expect.assertions(8);

		const promises = expectedArr.map( expected => {
			return new Promise( async resolve => {
				{
					// 1. The shelf retrieved with getBookShelf matches the expected one
					const shelf = component.getBookShelf(expected.id);
					expect(shelf).toBe(expected.shelf);
				}
				{
					// 2. Double checking that the stored book contains the expected shelf
					const book = await BooksAPI.get(expected.id);
					expect(book).toEqual(expect.objectContaining(expected));
				}
				resolve();
			});
		});

		await Promise.all(promises);
	});
});

