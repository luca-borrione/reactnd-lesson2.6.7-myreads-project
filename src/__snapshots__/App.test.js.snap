// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div
  className="list-books"
>
  <div
    className="list-books-title"
  >
    <h1>
      MyReads
    </h1>
  </div>
  <div
    className="list-books-content"
  >
    ;
  </div>
  <div
    className="open-search"
  >
    <a
      href="/search"
      id="goto-search"
      onClick={[Function]}
    >
      Add a book
    </a>
  </div>
</div>
`;
